Editorial: Tweak the parameter list of Function constructors

jmdyck opened the pull request:
In the specification of each of the constructors:
- Function
- GeneratorFunction
- AsyncGeneratorFunction
- AsyncFunction

the parameter list is given as
`( _p1_, _p2_, &hellip; , _pn_, _body_ )`

Note that these specific parameters are not referenced by the preamble or the algorithm. The preamble used to refer to them, before PR #2592; I don't think the algorithm has ever referred to them. Instead, the algorithm currently says:
```
Let _args_ be the _argumentsList_ that was passed
   to this function by [[Call]] or [[Construct]].
```

The wording of this step is problematic:
- It refers to an alias that is defined in another algorithm (which is prompting some special-casing in PR #2901).
- It says that `[[Call]]` or `[[Construct]]` is passing something to this function, which is incongruous. (`[[Call]]` and `[[Construct]]` deal with an arguments list that has already *been* passed to the function.)

To resolve all of the above, this PR changes the parameter list to:
`( ..._args_ )`
and just deletes the `Let _args_ be ...` step.

ljharb commented on spec.html:
```diff
@@ -29912,12 +29912,11 @@ <h1>The Function Constructor</h1>
 
       <emu-clause id="sec-function-p1-p2-pn-body">
-        <h1>Function ( _p1_, _p2_, &hellip; , _pn_, _body_ )</h1>
+        <h1>Function ( ..._args_ )</h1>
```
Function has a length of 1, and this syntax would give it a length of 0 (same for all 4)

jmdyck replied:
Ah, good point.

ljharb replied:
ah, true. it would still be nice if we could remove those sections by making the signatures `(arg1, ...rest)` or something.

jmdyck replied:
Each of the 4 constructors has a section that says the value of the `length` property is 1. E.g. [20.2.2.1 Function.length](https://tc39.es/ecma262/#sec-function.length)

ljharb replied:
ah, true. it would still be nice if we could remove those sections by making the signatures `(arg1, ...rest)` or something.

jmdyck replied:
It would be nice, but I don't think it's possible here without making things worse. E.g., with that particular suggestion, we'd then have to:
- explain that passing 0 args is okay (and isn't equivalent to passing `*undefined*` to `arg1`);
- (perhaps) stress that it's the *last* arg that's distinctive, not the first; and
- say something like `If _arg1_ is absent then let _args_ be an empty List; else, let _args_ be the list-concatenation of « _arg1_ » and _rest_.`, or else go back to a more indirect way of defining `_args_`.

michaelficarra commented:
Why not `( ... _parameterNames_, _bodyText_ )`? I don't like how this current change is less descriptive.

jmdyck commented:
> Why not `( ... _parameterNames_, _bodyText_ )`?

Because that's a syntax error? Granted, the spec doesn't *have* to use valid syntax when specifying built-ins (witness the status quo for these constructors), but I think it's better if it does. Plus bullets 1 and 3 above.

> I don't like how this current change is less descriptive.

I agree it's a bit of a loss. But I think there's enough help to make up for it:
- Each of these constructors has a preamble that tells you how the arguments are used.
- `Function` has a Note with examples, which the other constructors link to.

See also the [Date](https://tc39.es/ecma262/#sec-date), [Array](https://tc39.es/ecma262/#sec-array), and [_TypedArray_](https://tc39.es/ecma262/#sec-typedarray) constructors, where we don't try to convey meaning in the parameter-list.

michaelficarra commented:
Discussed at editor call today and we prefer my suggestion of `( ... _parameterNames_, _bodyText_ )` for clarity.

jmdyck commented:
I added a fixup to switch to the more descriptive parameter list. However:
- I used `_parameterArgs_` instead of `_parameterNames_`, because the latter suggests that only names (identifiers) are allowed, whereas anything matching (the appropriate variant of) _FormalParameters_ is allowed.
- I used `_bodyArg_` instead of `_bodyText_`, because we usually reserve "text" for a sequence of Unicode code points (and also because CreateDynamicFunction refers to it as `_bodyArg_`).

Then I noticed that, now that we had that separation between parameter-args and body-arg, it was silly to construct a list of all the args, only for CreateDynamicFunction to have to deconstruct it. So I added another commit that modified CreateDynamicFunction to take `_parameterArgs_` and `_bodyArg_` directly.

jmdyck commented:
(Hah, ecmarkup is complaining about the malformed parameter lists.)

bakkot commented:
Yeah I'll need to update ecmarkup. Which unfortunately is going to either mean a backport (ugh) or landing https://github.com/tc39/ecma262/pull/2901 first after all and rebasing this. We'll talk about it at the call today.

jmdyck commented:
(force-pushed to squash it down to 1 commit)

michaelficarra commented:
I've asked the esmeta developers to add support for this new parameter list form.

jhnaldo commented:
> I've asked the esmeta developers to add support for this new parameter list form.

Hi @michaelficarra, I have tried to update ESMeta to support this new parameter list form.
I have several questions about it.

While the original _rest_ parameters have been used at the end of a parameter list, this new parameter list form is used at the beginning of a parameter list.

Then, can this new form be used in any place in a parameter list?

In addition, what is the exact algorithm of argument assignment to parameters?

I think arguments are assigned to normal parameters first and then, the remaining arguments are assigned to this new form:
```
Function(1)
- _parameterNames_ : []
- _bodyText_: 1
Function(1, 2)
- _parameterNames_ : [1]
- _bodyText_: 2
```
Would this rule be still valid in general?
For example, assume that I define a new built-in algorithm `A ( ... _a_, _b_, ... _c_, _d_ )`. Then, is the expected result as follows?
```
A(1)
- _a_: not assigned
- _b_: 1
Function(1, 2)
- _parameterNames_ : [1]
- _bodyText_: 2
```

jhnaldo commented:
> I've asked the esmeta developers to add support for this new parameter list form.

Hi @michaelficarra, I have tried to update ESMeta to support this new parameter list form.
I have several questions about it.

While the original _rest_ parameters have been used at the end of a parameter list, this new parameter list form is used at the beginning of a parameter list.

Then, can this new form be used in any place in a parameter list?

In addition, what is the exact algorithm of argument assignment to parameters?

I think arguments are assigned to normal parameters first and then, the remaining arguments are assigned to this new form:
```
Function(1)
- _parameterNames_ : []
- _bodyText_: 1
Function(1, 2)
- _parameterNames_ : [1]
- _bodyText_: 2
```
Would this rule be still valid in general?
For example, assume that I define a new built-in algorithm `A ( _a_, ... _b_, _c_ )`. Then, is the expected result as follows?
```
A(1)
- _a_ : 1
- _b_ : []
- _c_: not assigned
A(1, 2)
- _a_: 1
- _b_: []
- _c_: 2
A(1, 2, 3)
- _a_ : 1
- _b_ : [2]
- _d_ : 3
```

michaelficarra commented:
@jhnaldo Correct.
